keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -159,69 +159,48 @@ take an aligned approach to adding apply to clients in a typesafe way.
 The client-go typed clients will be extended to include Apply functions, e.g.:
 
 ```go
-func (c *deployments) Apply(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
-func (c *deployments) ApplyStatus(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
-func (c *deployments) ApplyScale(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
+func (c *deployments) ApplyStatus(ctx Context, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
+func (c *deployments) ApplyScale(ctx Context, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions) (*Deployment, error)
 ```
 
 `ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
-to be used over the wire. This will make it obvious in the Apply function
-signature that `fieldManager` is required.
+to be used over the wire:
 
 ```go
 type ApplyOptions struct {
   DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
   Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+  FieldManager string `json:"fieldManager,omitempty" protobuf:"bytes,3,name=fieldManager"`
 }
 
-func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+// NewApplyOptions constructs an ApplyOptions. fieldManager is the name
+// associated with the actor or entity that is performing the apply and
+// is must be non-empty. force should be set to true if the apply should
+// re-qcquire conflicting fields owned by other fieldManagers.
+func(o ApplyOptions) NewApplyOptions(fieldManager string, force bool)
 ```
 
-<<[UNRESOLVED @jpbetz ]>>
+`ApplyOptions` is introduced primarily to allow us to include the
+`NewApplyOptions` function, and also to future proof the API.
```

Comment:
Do we need the new type to create the new function? Plus it looks like the new type is public so nothing prevents people from creating it in the first place.

Your function doesn't return anything and modifies the object?

What about:
```go
func NewApplyOptions(fieldManager string, force bool) PatchOptions
```

Reply:
How does it future proof the API? Do you mean in case we end-up adding new Apply specific fields? If that's the case, I would expect that we would could do:
```go
type ApplyOptions struct {
    PatchOptions
 
    // Add future flags here:
    //AdditionalSpecificOption string   
}
```

Reply:
Yeah, we have the ability to add apply specific fields in the future.

Embedding instead of copying over the fields works. The reason I had the fields copied over is that I wanted to add godoc to each of them, since I don't feel that the godoc on PatchOptions is optimal for explaining how the options should be used with Apply.

We might also want to make Force a non-pointer bool.

Reply:
> Do we need the new type to create the new function? Plus it looks like the new type is public so nothing prevents people from creating it in the first place.
> 
> Your function doesn't return anything and modifies the object?
> 
> What about:
> 
> ```go
> func NewApplyOptions(fieldManager string, force bool) PatchOptions
> ```

Oops, I wrote that signature in a rush. What I wanted is:

```go
func NewApplyOptions(fieldManager string, force bool) ApplyOptions
```

Reply:
update: I'm gonna drop NewApplyOptions entirely and try to do everything by improving ApplyOptions.